shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java

```diff
@@ -134,6 +163,139 @@ public void sendAppPrivateCommand(String action, Bundle data) {
     restartAlwaysRequired = isRestartAlwaysRequired();
   }
 
+  // Loosely based off of
+  // https://github.com/android/user-interface-samples/blob/master/WindowInsetsAnimation/app/src/main/java/com/google/android/samples/insetsanimation/RootViewDeferringInsetsCallback.kt
+  //
+  // When the IME is shown or hidden, it immediately sends an onApplyWindowInsets call
```

Comment:
Ah, this is much better, thanks! Does this class also handle all the non-animating inset changes? Like the nav bar, status bar going in and out of fullscreen etc? How do all the phases like onStart, onApplyWindowInsets etc affect those insets?

Reply:
Non-animating changes are just passed right on through to FlutterView.onApplyWindowInsets without impact. The ime-type filtering also prevents this from doing anything on any non-ime animations. During animations, changes to the window are handled by the animation controller, which will end the current animation, and restart a new one with the modified settings to continue. For example, rotating the device mid-keyboard animation just causes the animation to continue where it left off but scaled to the rotated window.